flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/RegularOperatorChain.java

```diff
@@ -216,24 +248,64 @@ private OperatorSnapshotFutures buildOperatorSnapshotFutures(
         return snapshotInProgress;
     }
 
-    private static OperatorSnapshotFutures checkpointStreamOperator(
+    private OperatorSnapshotFutures checkpointStreamOperator(
             StreamOperator<?> op,
             CheckpointMetaData checkpointMetaData,
             CheckpointOptions checkpointOptions,
             CheckpointStreamFactory storageLocation,
             Supplier<Boolean> isRunning)
             throws Exception {
         try {
-            return op.snapshotState(
-                    checkpointMetaData.getCheckpointId(),
-                    checkpointMetaData.getTimestamp(),
-                    checkpointOptions,
-                    storageLocation);
+            if (operatorEventDispatcher.containsOperatorEventGateway(op.getOperatorID())) {
+                operatorEventDispatcher.snapshotOperatorEventGatewayState(
+                        op.getOperatorID(), getOperatorStateBackend(op));
+            }
+
+            OperatorSnapshotFutures futures =
+                    op.snapshotState(
+                            checkpointMetaData.getCheckpointId(),
+                            checkpointMetaData.getTimestamp(),
+                            checkpointOptions,
+                            storageLocation);
+
+            if (operatorEventDispatcher.containsOperatorEventGateway(op.getOperatorID())) {
+                operatorEventDispatcher.notifyOperatorSnapshotStateCompleted(
```

Comment:
Should we invoke this method only after all `futures` are successfully completed?

What is the expected behavior if `OperatorSnapshotFutures` fails?

Reply:
Although `OperatorSnapshotFutures` have not been completed when the method above returns, it is guaranteed that the states to be saved into the snapshot would not be changed by any following operations on the operator, thus it is safe to send ACK events from the operator to its OC and release the buffered events.

According to our offline discussion, I'll update the closing gateway process, to close gateways before sending `AcknowledgeCloseGatewayEvent`, so as to make sure there will be no event reaching the OC after the `AcknowledgeCloseGatewayEvent`.